experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,565 @@
+//! pub trait Trait {
+//!     type IterItem;
+//!     fn variants() -> Box<dyn ExactSizeIterator<Item=Self::IterItem>>;
+//! }
+//! ```
+//!
+//! Getting to a generic iterator that doesn't require a box is a piece of gymnastics.  Getting
+//! to an iterable of owned strings requires the iterating trait to be implemented over
+//! *a lifetime-scoped reference* of the implementing type.
+//!
+//! ```rust
+//! pub trait Variants {
+//!     /// The type of the item yieled by the iterator returned by [Variants::variants].  Note
+//!     /// that [Variants::Variant] may be a reference to the type stored in the iterator.
+//!     type Variant;
+//!     /// The type of the iterator returned by [Variants::variants].
+//!     type Iter: ExactSizeIterator<Item = Self::Variant>;
+//!     fn variants(self) -> Self::Iter;
+//! }
+//!
+//! // Here's how to implement the trait when the underlying type is borrowed.
+//!
+//! pub struct BorrowedVariant {
+//!     variants: Vec<&'static str>,
+//! }
+//!
+//! impl<'a> Variants for &'a BorrowedVariant {
+//!     type Variant = &'a &'a str;
+//!     type Iter = std::slice::Iter<'a, &'a str>;
+//!     fn variants(self) -> Self::Iter {
+//!       self.variants.iter()
+//!     }
+//! }
+//!
+//! let borrowed = BorrowedVariant{ variants: vec!["a", "b"], };
+//! assert_eq!(
+//!     vec!["a", "b"],
+//!     borrowed.variants()
+//!         .map(|v| v.to_owned().to_owned())
+//!         .collect::<Vec<String>>(),
+//! );
+//!
+//! // Here is how to implement the trait when the underlying type is owned.
+//!
+//! pub struct OwnedVariant {
+//!     variants: Vec<String>,
+//! }
+//!
+//! impl<'a> Variants for &'a OwnedVariant {
+//!     type Variant = &'a String;
+//!     type Iter = std::slice::Iter<'a, String>;
+//!     fn variants(self) -> Self::Iter {
+//!         self.variants.iter()
+//!     }
+//! }
+//!
+//! let owned = OwnedVariant{ variants: vec!["a".to_string(), "b".to_string()], };
+//! assert_eq!(
+//!     vec!["a", "b"],
+//!     owned.variants()
+//!         .map(|v| v.to_owned())
+//!         .collect::<Vec<String>>(),
+//! );
+//!
+//! ```
+//!
+
+/// This trait contains the common features of the Locale object that must be shared among
+/// all the implementations.  Every implementor of `listformat` should provide their
+/// own version of [Locale], and should ensure that it implements [Locale]. as
+/// specified here.
+///
+/// A stronger type definition is also possible, pending agreement.  Until no firm implementations
+/// exist, we reserve the option to strenghten the contract of [Locale] as we see fit.
+///
+/// For example, here are some possible methods that this trait could implement instead:
+///
+/// - `language`
+/// - `script`
+/// - `variant`
+/// - (...)
+pub trait Locale<'a> {
+    /// Gives the full representation of the `Locale` in the [BCP 47][bcp47] string encoding.
+    ///
+    ///   [bcp47]: https://tools.ietf.org/html/bcp47
+    fn to_bcp47() -> &'a str;
+}
```

Comment:
You don't need a lifetime here. `Locale` cannot freely be presented as `str`, so we need `to_string() -> String`.
`as_str` can only be used for cases where the memory can be presented as a string slice, so individual subtags can mostly do that, but a combination of them needs to have `to_bcp47_string`.

Reply:
Turned this entire trait into one lower-bounded by `std::fmt::Display` as agreed. 
Added a section in the text about this change.